types/index.d.ts

```diff
@@ -2084,13 +2104,30 @@ declare module 'mongoose' {
   type TreatAsPrimitives = actualPrimitives |
   NativeDate | RegExp | symbol | Error | BigInt | Types.ObjectId;
 
+  // This will -- when possible -- extract the original type of the subdocument in question
+  type LeanSubdocument<T> = T extends (Types.Subdocument<Require_id<T>['_id']> & infer U) ? LeanDocument<U> : Omit<LeanDocument<T>, '$isSingleNested' | 'ownerDocument' | 'parent'>;
+
   type LeanType<T> =
     0 extends (1 & T) ? T : // any
       T extends TreatAsPrimitives ? T : // primitives
-        T extends Types.Subdocument ? Omit<LeanDocument<T>, '$isSingleNested' | 'ownerDocument' | 'parent'> : // subdocs
+        T extends Types.Subdocument ? LeanSubdocument<T> : // subdocs
           LeanDocument<T>; // Documents and everything else
 
-  type LeanArray<T extends unknown[]> = T extends unknown[][] ? LeanArray<T[number]>[] : LeanType<T[number]>[];
+  // Used only when collapsing lean arrays for ts performance reasons:
+  type LeanTypeOrArray<T> = T extends unknown[] ? LeanArray<T> : LeanType<T>;
+
+  type LeanArray<T extends unknown[]> =
+  // By checking if it extends Types.Array we can get the original base type before collapsing down,
+  // rather than trying to manually remove the old types. This matches both Array and DocumentArray
+    T extends Types.Array<infer U> ? LeanTypeOrArray<U>[] :
+      // If it isn't a custom mongoose type we fall back to "do our best"
+      T extends unknown[][] ? LeanArray<T[number]>[] : LeanType<T[number]>[];
+
+  type _WithoutFunctions<T extends {}> = {
+    [K in keyof T]: [0] extends [1 & T[K]] ? K : // keep any
+      T[K] extends Function ? never : K;
+  };
+  type StripFunctions<T extends {}> = Pick<T, _WithoutFunctions<T>[keyof T]>;
```

Comment:
This change is a problem. We actually had to get rid of similar logic because of https://github.com/microsoft/TypeScript/issues/43281 . There's no way to reliably "remove just function properties" from a TypeScript type, particularly when generics are involved.

Reply:
Interesting; I'll have to dig into that issue. It still seems like it should be possible to strip them out only if we can be sure they are functions, but from a brief look at the issue what you were doing seems like what is wanted, so I'm likely missing something.

I'll either remove this or else prove a fix -- possibly in another PR.

Reply:
I've removed this for now; I will revisit in another PR